server.go

```diff
@@ -183,74 +190,41 @@ func (s *SockchatServer) editProfile(w http.ResponseWriter, r *http.Request) {
 
 }
 
-func (s *SockchatServer) shutConnection(conn *SockChatWS) {
-	s.store.DisconnectUser(conn)
-	conn.Close()
-}
-
-func (s *SockchatServer) sendMessageToChannel(request SocketMessage, conn *SockChatWS) {
-	req := SendMessageRequest{}
-	if err := json.Unmarshal(request.Payload, &req); err != nil {
-		log.Printf("error while unmarshaling request for sending message: %v", err)
-		return
-	}
-	if !s.store.ChannelHasUser(req.Channel, conn) {
-		conn.WriteSocketMsg(NewErrorMessage("you are not member of this channel"))
-		return
-	}
-	channel, _ := s.store.GetChannel(req.Channel)
-	for user := range channel.Users {
-		user.WriteSocketMsg(NewSocketMessage("new_message", MessageEvent{Channel: req.Channel, Author: conn.nick, Text: req.Text}))
+func (s *SockchatServer) authorizeConnection(ctx context.Context, request SocketMessage, conn *SockChatWS) error {
+	// user already logged in
+	if conn.handler != nil {
+		return nil
 	}
 
-}
-
-func (s *SockchatServer) loginUser(ctx context.Context, request SocketMessage, conn *SockChatWS) {
-	req := UnmarshalLoginRequest(request.Payload)
-	if err := s.userService.LoginUser(ctx, req.Nick, req.Password); err != nil {
-		conn.WriteSocketMsg(NewErrorMessage(err.Error()))
-	} else {
-		conn.nick = req.Nick
-		conn.WriteSocketMsg(NewSocketMessage(fmt.Sprintf("logged_in:%s", req.Nick), "{}"))
-		conn.SetReadDeadline(time.Now().Add(s.timeoutAuthorized))
+	if request.Action != LoginAction {
+		return fmt.Errorf("you must log in first using " + LoginAction + " action")
 	}
-}
-
-func (s *SockchatServer) createNewChannel(request SocketMessage, conn *SockChatWS) {
-	channel := UnmarshalChannelRequest(request.Payload)
-	if err := s.store.CreateChannel(channel.Name); err != nil {
-		conn.WriteSocketMsg(NewErrorMessage(err.Error()))
-	} else {
-		conn.WriteSocketMsg(NewSocketMessage("channel_created", channel))
+	req := UnmarshalLoginRequest(request.Payload)
+	if !s.userProfiles.IsAuthValid(ctx, req.Nick, req.Password) {
+		return fmt.Errorf("login rejected: invalid credentials")
 	}
-}
-
-func (s *SockchatServer) joinChannel(request SocketMessage, conn *SockChatWS) {
-	channel := UnmarshalChannelRequest(request.Payload)
-	if err := s.store.AddUserToChannel(channel.Name, conn); err != nil {
-		conn.WriteSocketMsg(NewErrorMessage(err.Error()))
+	if s.userHandlers[req.Nick] != nil {
+		conn.handler = s.userHandlers[req.Nick]
 	} else {
-		conn.WriteSocketMsg(NewSocketMessage("channel_joined", ChannelUserChangeEvent{Channel: channel.Name, Nick: conn.nick}))
+		conn.handler = &UserHandler{Nick: req.Nick, connections: make(map[*SockChatWS]bool), requests: make(chan *UserAction)}
```

Comment:
```suggestion
		conn.handler = &UserHandler{Nick: req.Nick, connections: make(map[*SockChatWS]bool), requests: make(chan *UserAction, 10)}
```
Consider using buffered channel for this use case: https://go.dev/tour/concurrency/3